### PR TITLE
Pecos hybrid channel testing

### PR DESCRIPTION
**Pull request by clarkpede**
This is PR 3 of a series of PRs intended to get the pecos-dev branch back in line with the "best" and most stable code we are currently using.

### CODE REVIEW

File: SU2_CFD/include/hybrid_RANS_LES_forcing.inl [hunk 1]
```diff
@@ -91,9 +92,22 @@ inline void CHybridForcingTG0::SetTGField(
...
+  //h[2] = C * sin(a[0]*x[0]) * sin(a[1]*(x[1]-1.0)) * cos(a[2]*x[2]-M_PI/2);
+  // // CHANNEL HACK
+  // h[0] = A * cos(a[0]*x[0]) * sin(a[1]*(x[1]-1.0)) * sin(a[2]*x[2]);
+  // h[1] = B * sin(a[0]*x[0]) * cos(a[1]*(x[1]-1.0)) * sin(a[2]*x[2]);
+  // h[2] = C * sin(a[0]*x[0]) * sin(a[1]*(x[1]-1.0)) * cos(a[2]*x[2]);
+
+  // CHANNEL HACK
+  h[0] = A * cos(a[0]*x[0]) * sin(a[1]*x[1]) * sin(a[2]*(x[2]-1.0));
+  h[1] = B * sin(a[0]*x[0]) * cos(a[1]*x[1]) * sin(a[2]*(x[2]-1.0));
+  h[2] = C * sin(a[0]*x[0]) * sin(a[1]*x[1]) * cos(a[2]*(x[2]-1.0));
```
Comment (clarkpede) [inline SU2_CFD/include/hybrid_RANS_LES_forcing.inl]: @trevilo Why is this shift (`x[2] - 1`) necessary?

File: SU2_CFD/src/fluctuating_stress.cpp [hunk 2]
```diff
@@ -111,41 +111,90 @@ void CM43Model::CalculateEddyViscosity(const CGeometry* geometry,
...
 
   for (unsigned short iDim = 0; iDim < nDim; iDim++) {
     for (unsigned short jDim = 0; jDim < nDim; jDim++) {
-      eddy_viscosity[iDim][jDim] =
-          density * C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+      eddy_viscosity[iDim][jDim] = density * C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+      //C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+	//C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+	//density * C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+	//density * damping*C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
```
Comment (clarkpede) [inline SU2_CFD/src/fluctuating_stress.cpp]: @trevilo We calculate damping here, but we don't use it? What's the status on the damping?

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp [hunk 3]
```diff
@@ -351,21 +352,22 @@ void CHybrid_Mediator::SetupResolvedFlowSolver(const CGeometry* geometry,
...
+    //   solver_container[FLOW_SOL]->average_node[iPoint]->GetResolutionAdequacy();
+    // assert(avg_resolution_adequacy >= 0);
+    // assert(avg_resolution_adequacy == avg_resolution_adequacy);
+    // //const su2double factor = pow(min(avg_resolution_adequacy, 10.0), 4.0/3);
+    // const su2double factor = max(min(avg_resolution_adequacy, 10.0), 1.0);
+    // for (unsigned short iDim = 0; iDim < nDim; iDim++) {
+    //   for (unsigned short jDim = 0; jDim < nDim; jDim++) {
+    //      aniso_eddy_viscosity[iDim][jDim] *= factor;
+    //   }
+    // }
```
Comment (clarkpede) [inline SU2_CFD/src/hybrid_RANS_LES_model.cpp:283]: @trevilo Is there a reason this was removed?  Was it also removed in CDP?

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp [hunk 4]
```diff
@@ -534,6 +536,15 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
...
 
+  // Make tauSGET trace = 0
+  su2double traceTauSGET = 0;
+  for (iDim =0; iDim < nDim; iDim++) {
+    traceTauSGET += tauSGET[iDim][iDim];
+  }
+  for (iDim = 0; iDim < nDim; iDim++) {
+    tauSGET[iDim][iDim] -= traceTauSGET/3;
+  }
+
```
Comment (clarkpede) [inline SU2_CFD/src/hybrid_RANS_LES_model.cpp:460]: @trevilo Shouldn't we subtract out the trace of tauSGET where it's defined in the CAvgGrad_Hybrid as well?

In the long term, we need to find a way to eliminate the duplication between numerics and the InvLengthTensor calculation.  I've got an idea or two, but I haven't had time to work on them.

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp [hunk 5]
```diff
@@ -584,10 +599,12 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
       for (jDim = 0; jDim < nDim; jDim++) {
         for (kDim = 0; kDim < nDim; kDim++) {
           // rotation contribution
-          Pij[iDim][jDim] += 2.0*alpha*eddy_viscosity*Sd_avg[iDim][kDim]*Om[kDim][jDim];
+          Pij[iDim][jDim] += 2.0*alpha_fac*eddy_viscosity*Sd_avg[iDim][kDim]*Om[kDim][jDim];
         }
         // rho*k contribtuion
-        Pij[iDim][jDim] -= 2.0*alpha*rho*ktot*(Sd[iDim][jDim]+Om[iDim][jDim])/3.0;
+	// should this part be just alpha scaling?
```
Comment (clarkpede) [inline SU2_CFD/src/hybrid_RANS_LES_model.cpp]: Yes, yes it should.

File: SU2_CFD/src/solver_direct_mean.cpp [hunk 6]
```diff
@@ -17702,24 +17715,36 @@ void CNSSolver::Source_Residual(CGeometry *geometry, CSolver **solver_container,
...
       // energy
       Residual[nDim+1] = 0.0;
-      for (unsigned short iDim = 0; iDim < nDim; iDim++)
-        Residual[nDim+1] += -Volume * U[iDim+1] * Force[iDim];
+      for (unsigned short iDim = 0; iDim < nDim; iDim++) {
+        //Residual[nDim+1] += -Volume * U[iDim+1] * Force[iDim];
+	//Residual[nDim+1] += -Volume * U[iDim+1] * (Force[iDim] - MeanForce[iDim]);
+	Residual[nDim+1] += 0.0; // no source in energy?
+	//Residual[nDim+1] += -Volume * (U[iDim+1]/U[0]) * (Force[iDim] - MeanForce[iDim]);
+      }
```
Comment (clarkpede) [inline SU2_CFD/src/solver_direct_mean.cpp:17745]: @trevilo What's going on here?

File: SU2_CFD/include/hybrid_RANS_LES_forcing.inl (hunk 1 above)
Comment (trevilo) [reply SU2_CFD/include/hybrid_RANS_LES_forcing.inl]: I don't think it is necessary.  I originally shifted z (my wall-normal coordinate) b/c Sigfried had shifted y (his wall-normal coordinate) in his first testing in CDP.  He did this for symmetry purposes, but later stopped b/c it didn't seem to matter much (or that is my understanding at least).  In any case, I don't think this shift combined with the form here actually gives the right symmetry.  It is symmetric in the wall-normal component when you would want that to be anti-symmetric, right?  This is true in CDP as well.  So, bottom line is that I think this can be safely eliminated.

File: SU2_CFD/src/fluctuating_stress.cpp (hunk 2 above)
Comment (trevilo) [reply SU2_CFD/src/fluctuating_stress.cpp]: Yeah, the damping or wall-blending of the SGET term is something I messed around with, but without ever coming to a definitive conclusion for how it should be handled.  You can see evidence of this in the code just below, where as similar blending is applied as what is in CDP (although it blends to 0 rather than the RANS eddy viscosity b/c I had stability problems when blending to the RANS eddy viscosity).

In any case, Sigfried did not need this blending for the channel in CDP.  He added it later for very high aspect ratio cells in the WMH.  Empirically, I observed that we did need something (for stability) in SU2, and the stability of the code was somewhat sensitive to how exactly it was done.  However, those observations are mostly from when the SGET term was treated fully explicitly.  So, it should be less sensitive now I would hope.

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp (hunk 3 above)
Comment (trevilo) [reply SU2_CFD/src/hybrid_RANS_LES_model.cpp:283]: No, it is still in CDP.  I commented it out for the channel runs for two reasons: 1) it is hacky so I just wanted to see if we really needed it (at least for the channel), and 2) for stability again.  Now, as before, the stability issues were observed prior to any implicitness on the SGET term.  So, I was trying to keep mu^{SGET} from getting big.  That shouldn't really be a concern anymore, but I have not tested with this back on.

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp (hunk 5 above)
Comment (trevilo) [reply SU2_CFD/src/hybrid_RANS_LES_model.cpp]: I agree, which is why the code is that way.  But, just food for thought, Sigfried says he gets better results on the periodic hill when the whole stress is scaled by a*(2-a), not just the anisotropic part. Given the definition of alpha, scaling by a*(2-a) on the isotropic part is clearly an inconsistency.  So... the fact that that works better must mean it is covering for some other issue.

File: SU2_CFD/src/solver_direct_mean.cpp (hunk 6 above)
Comment (trevilo) [reply SU2_CFD/src/solver_direct_mean.cpp:17745]: Ha, just me being dumb.  I originally had the forcing entering the total energy equation, since that is what you'd do for a "normal" forcing term (i.e., a source in the momentum equation).  But, the forcing here is a transfer from unresolved to resolved.  Therefore, there isn't any forcing contribution to the total energy equation.

File: SU2_CFD/include/hybrid_RANS_LES_forcing.inl (hunk 1 above)
Comment (clarkpede) [reply SU2_CFD/include/hybrid_RANS_LES_forcing.inl]: Alright.  I'll compare a short run both with and without the shift.

File: SU2_CFD/src/fluctuating_stress.cpp (hunk 2 above)
Comment (clarkpede) [reply SU2_CFD/src/fluctuating_stress.cpp]: Alright.  I'm moving this to a config option, with three choices:

1. No high-AR damping
2. Blend to zero for high-AR
3. Blend to RANS for high-AR

File: SU2_CFD/src/solver_structure.cpp [hunk 7]
```diff
@@ -3025,7 +3025,8 @@ void CSolver::UpdateAverage(const su2double weight, const unsigned long iPoint,
   const su2double* current = node[iPoint]->GetSolution();
 
   for (unsigned short iVar = 0; iVar < nVar; iVar++) {
-    const su2double new_average = (current[iVar] - average[iVar])*weight + average[iVar];
+    // BE for everything
+    const su2double new_average = (current[iVar]*weight + average[iVar])/(1.0 + weight);
```
Comment (clarkpede) [inline SU2_CFD/src/solver_structure.cpp]: If we use BE here, we should use BE everywhere (including for our hybrid variables like production).  Right now CNSSolver::UpdateAverage still uses FE.  I can add a fix to change the weight to be the "weight" for BE, which will make everything consistent again.

I don't know if this will change things much, but we should at least be consistent.

**Comment by clarkpede** [comment]
So there is a slight change after ~6 flow-throughs (3000 timesteps) with my recent "clean-up" commits.  I tracked the change down to commit f074f3325c0834b8c84b77f963d168e7cc2d016d.  I changed the boundary conditions in v2f to fix what I think was a bug.

I'll do a longer run to see how much things differ over 7000 timesteps (~13 flow-throughs).  If we're still happy with the results, that will be a second "baseline" against which I'll compare future changes.

File: SU2_CFD/src/solver_structure.cpp (hunk 7 above)
Comment (trevilo) [reply SU2_CFD/src/solver_structure.cpp]: I agree: they should all be the same.  At one point early on when I was having stability problems with the channel I changed them all to BE, just to see if it helped.  It didn't, and I don't think it should affect anything much one way or the other, but I guess I forgot to switch these back.  Anyway, we can go back to FE or add an input option.  Whatever you prefer.

**Comment by trevilo** [comment]
> So there is a slight change after ~6 flow-throughs (3000 timesteps) with my recent "clean-up" commits. I tracked the change down to commit [f074f33](https://github.com/pecos-hybrid/SU2/commit/f074f3325c0834b8c84b77f963d168e7cc2d016d). I changed the boundary conditions in v2f to fix what I think was a bug.

Regarding that BC fix, I agree there was a bug in using the instantaneous flow rather than the mean.  I have a similar mod on the axibump testing branch.  I used iPoint rather than jPoint (on the wall rather than 1 off) for the density and viscosity.  I'm not sure off the top of my head which it should actually be, so... just something to check.

**Comment by clarkpede** [comment]
> I used iPoint rather than jPoint (on the wall rather than 1 off) for the density and viscosity. I'm not sure off the top of my head which it should actually be, so... just something to check.

I used the first off-node value because that's what's done in SST.  I did check it just now, though.  [Kalitzin](https://arc.aiaa.org/doi/pdf/10.2514/6.1999-3780), explicitly stated that he used the first point off the wall for his values of k and y that appear in the boundary conditions.  Unfortunately, that paper was incompressible, so viscosity and density were uniform.  The only transonic v2-f paper I know doesn't specify which point was used.  The NASA TMR specifies the first off-wall node.

The first off-wall node makes sense to me.  If I understand the BC correctly, we're trying to indirectly enforce a derivative boundary condition.  Therefore, we need the values at the first off-wall node and the values at the wall.  But since k_w = 0, the wall terms disappear, and we're only left with the first node terms.  Please correct me if I'm wrong.

**Comment by clarkpede** [comment]
So here's the difference after 7000 timesteps.  The picture is very similar for 6000 and 6500 timesteps.  These are all calculated with planar averages, which means "u" is the planar average of the instantaneous velocity.

Long_time_PR_v2f

`PR_1` is commit f074f33 and baseline is the `channel_lm` tag.  They both look "right".  Having a higher alpha makes sense to me, since we're damping fluctuations by switching from instantaneous values to average values in the boundary conditions.

**Comment by trevilo** [comment]
> I used the first off-node value because that's what's done in SST.

Ok, this seems like a good enough reason.  I think it actually doesn't really matter which we use (see below).

> The first off-wall node makes sense to me. If I understand the BC correctly, we're trying to indirectly enforce a derivative boundary condition. 

Yes, we're trying to get the right behavior of k, specifically k=0, which is set directly, and dk/dy=0, which is consistent with setting the right value of epsilon, namely \varepsilon(0) = \lim_{y \rightarrow 0} 2 \nu k / y^2.  

> Therefore, we need the values at the first off-wall node and the values at the wall. But since k_w = 0, the wall terms disappear, and we're only left with the first node terms. Please correct me if I'm wrong.

I don't entirely follow you here.  We need to approximate the limit above.  Since k and y are zero at the wall, you can't just evaluate it, and so we approximate it with data from the first point off the wall.  In incompressible solvers you get k and y at the first node off the wall, and since \nu is constant there is no ambiguity.  When \nu varies we have to make a choice, either evaluate the whole expression at the first node off the wall, or take \nu at the wall and k and y from the first node off.  Both are consistent approximations that will approach the true limit as the first point gets closer to the wall, so it isn't immediately obvious to me that we should prefer one over the other.

I kind of prefer to use \nu at the wall, just because there isn't any reason to approximate that part of the limit expression, but... that isn't really a strong reason, and I doubt it matters much either way.  So, I'm happy to leave it as you have it.

**Comment by trevilo** [comment]
> `PR_1` is commit [f074f33](https://github.com/pecos-hybrid/SU2/commit/f074f3325c0834b8c84b77f963d168e7cc2d016d) and baseline is the `channel_lm` tag. They both look "right". Having a higher alpha makes sense to me, since we're damping fluctuations by switching from instantaneous values to average values in the boundary conditions.

Looks good.  I'm a tiny bit surprised it affects \alpha this much, but... fine.  I don't see any reason not to think it is right.

File: SU2_CFD/include/hybrid_RANS_LES_forcing.inl (hunk 1 above)
Comment (clarkpede) [reply SU2_CFD/include/hybrid_RANS_LES_forcing.inl]: The results are essentially the same:
Figure_6

Comment (trevilo) [reply SU2_CFD/include/hybrid_RANS_LES_forcing.inl]: Looks good.

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp (hunk 4 above)
Comment (trevilo) [reply SU2_CFD/src/hybrid_RANS_LES_model.cpp:460]: Oops, I think I missed this one the first time through.

Anyway... yeah, it should be consistent.  I don't really know that trace should be subtracted out in either spot.  In the model, the trace of tauSGET contributes to the modeled stress in the same way as (2/3)*alpha*k, so it at least ought to average to zero to avoid contaminating k.  The easiest way to achieve that is simply to have the trace be zero.  I'm pretty sure it was there at one point in CAvgGrad_Hybrid, b/c I remember commenting it out while investigating stability problems.

Bottom line: I'm fine with including it in CAvgGrad_Hybrid too, assuming stability is ok.

**Review by trevilo — COMMENTED** [review]
@clarkpede do you have channel results with this mod (i.e., 5779003) or is that the case sitting in the queue?

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp [hunk 8]
```diff
@@ -561,6 +557,11 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
   // 2) tauSGRS contribution.  NB: Neglecting divergence contribution
   // here.  TODO: Add divergence contribution.
 
+  /*--- Testing on the WMH indicates that scaling the whole stress by
+   * alpha*(2-alpha) improves the model performance.  That change would
+   * make the turbulent kinetic energy inconsistent, so it is avoided here.
+   * But that indicates there's some other issue. ---*/
+
   su2double alpha_fac = alpha*(2.0 - alpha);
```
Comment (trevilo) [inline SU2_CFD/src/hybrid_RANS_LES_model.cpp:482]: An update on this.  Sigfried indicates it isn't actually true.  I guess that conclusion (about scaling the whole stress by a*(2-a)) was based on looking at just one snapshot.  Anyway, now he says, "After many more runs, I was wrong about the scaling of the trace used in rM.  The best results come from alpha scaling on the isotropic part".

**Comment by clarkpede** [comment]
Alright.  After reverting the tauSGET trace, we're back to a relatively normal run.  It looks like changing everything to BE did change the behavior a little, but we're still in the same ballpark.  Here's some comparisons between the original state (the `channel_lm` tag) and the current state:
hybrid_26000
hybrid_27000

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp [hunk 9]
```diff
@@ -580,18 +497,21 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
...
+     * negative eigenvalues could result, which would be interpreted as
+     * "negative" lengthscales. ---*/
     for (iDim = 0; iDim < nDim; iDim++) {
       for (jDim = 0; jDim < nDim; jDim++) {
         for (kDim = 0; kDim < nDim; kDim++) {
-          Pij[iDim][jDim] += 2.0*alpha_fac*eddy_viscosity*Sd_avg[iDim][kDim]*Om[jDim][kDim];
+          // rotation contribution
+          Pij[iDim][jDim] += 2.0*alpha_fac*eddy_viscosity*Sd_avg[iDim][kDim]*Om[kDim][jDim];
+          // Isotropic contribution (from delta_ij div_vel)
+          Pij[iDim][jDim] += 2.0*alpha_fac*eddy_viscosity*Sd_avg[iDim][jDim]*div_avg_vel/3.0;
```
Comment (trevilo) [inline SU2_CFD/src/hybrid_RANS_LES_model.cpp]: Shouldn't this line use the divergence of the resolved velocity, not the average?

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp [hunk 10]
```diff
@@ -580,18 +497,21 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
...
         }
-        // TODO: Add rho*k contribution
+        // rho*k contribtuion
+	Pij[iDim][jDim] -= 2.0*alpha*rho*ktot*(Sd[iDim][jDim]+Om[iDim][jDim])/3.0;
```
Comment (trevilo) [inline SU2_CFD/src/hybrid_RANS_LES_model.cpp]: For RDELTA_INDICATOR_FULLP_WITH_DIV, isn't there a div(u) contribution to this term?

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp (hunk 9 above)
Comment (clarkpede) [reply SU2_CFD/src/hybrid_RANS_LES_model.cpp]: You're correct.  I'll fix that.

File: SU2_CFD/src/hybrid_RANS_LES_model.cpp (hunk 10 above)
Comment (clarkpede) [reply SU2_CFD/src/hybrid_RANS_LES_model.cpp]: Good catch.  I'll fix that.
